### PR TITLE
feat: add weth subparser

### LINKED ISSUE #389: Add new ethereum subParser to interpret weth deposit & withdrawal events as transfers
### Overview

Deposits to the wrapped ethereum contract do not emit erc20 "transfer" events, preventing us from properly tracking the balance change.

We need to add an ethereum weth subParser to capture these "deposit" and "withdrawal" events and interpret them as sends/receives erc20 transfers in the transacation.

We have an existing pattern for parsing transaction data for a given contracts abi that should be relatively straightforward

Basically we need to add a sub parser that returns a `data` and `transfers` field that we pass in here:
https://github.com/shapeshift/unchained/blob/develop/node/packages/client/src/ethereum/parser/index.ts#L70
https://github.com/shapeshift/unchained/blob/develop/node/packages/client/src/ethereum/parser/index.ts#L72

### Reference

Original discussion about this issue:
https://github.com/shapeshift/web/issues/1363

WETH contract:
https://etherscan.io/address/0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2

Withdrawal event example:
https://etherscan.io/tx/0x59ad10aae5c99b1d39df686208446d23b286ccd6bbc6bcd1f873b0c973cda840#eventlog

Deposit event example:
https://etherscan.io/tx/0x34dd6e7f0621d1083bd2950953f724a8d6ec8688ea440b0eb2de3d4f69c55d2b#eventlog

The new sub parser will need to return `transfers` and `data` in contractParserResult so that the parsedTransaction is populated on these lines:
https://github.com/shapeshift/unchained/blob/develop/node/packages/client/src/ethereum/parser/index.ts#L70
https://github.com/shapeshift/unchained/blob/develop/node/packages/client/src/ethereum/parser/index.ts#L72

Example of existing thorchain ethereum subParser being initialized:
https://github.com/shapeshift/unchained/blob/develop/node/packages/client/src/ethereum/parser/index.ts#L34

We use this file as a playground for testing many new changes:

https://github.com/shapeshift/lib/blob/main/packages/chain-adapters/src/ChainAdapterCLI.ts

One way of testing these changes would be:

1) Point to your locally running unchained here: https://github.com/shapeshift/lib/blob/main/packages/chain-adapters/src/ChainAdapterCLI.ts#L30

2) Uncomment relevant code to subscribe to local unchained eth websocket: https://github.com/shapeshift/lib/blob/main/packages/chain-adapters/src/ChainAdapterCLI.ts#L108

3) comment out code related to other chains (cosmos towards the bottom of file)

4) Add your wallet mnemonic to CLI_MNEMONIC in .env file

4) yarn adapterCli

At this point all of your prior eth transactions should print to the terminal and you can verify that they are being parsed correctly

### Acceptance Criteria

- A new weth subParser is created and initialized in a similar way to the other parsers here:
https://github.com/shapeshift/unchained/blob/develop/node/packages/client/src/ethereum/parser/index.ts#L38

- The `transfer` field should appear as a weth erc20 send for deposits and receive for withdrawals

- The `data` field should be { "parser" Txparser.weth, "method": "Deposit" } or  { "parser" Txparser.weth, "method": "Withdraw" }.

- Unit tests with realistic mock data should be included.

### Need By Date

_No response_

### Ownership

- [X] If my bounty needs engineering or needs product I have added the respective labels on the right
- [X] As the sponsor of this bounty I will review the changes in a preview environment (ops/product) or review the PR (engineering)

### Estimated effort

12 hours

### Sponsor / Stakeholder

elmutt (shapeshift discord & github)

### Bounty Hunters

- [Join our discord](https://discord.gg/shapeshift)
- Include an expected timeline for you to complete work in the work plan when you apply for this bounty!
- Please refer to [this link](https://shapeshiftdao.gitbook.io/getting-started/how-to-contribute/bounties#resources-for-new-contributors-and-bounty-hunters) for some basic info
- Please ***do not*** start work on this issue until you are approved in Gitcoin.

### PR DESCRIPTION
**Pull request by Soptq**
This PR adds WETH parser to parse `deposit()` and `withdraw(wad)`

close #389

### CODE REVIEW

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Nice one @Soptq, glad to see you back in our codebase!

Could you please add tests for the new parser in `ethereum.test.ts`? You'll need to add 2 additional transactions, a deposit and withdraw, that go through this parser to the `mockData` directory.

**Comment by elmutt** [comment]
Thanks @Soptq this looks great! It looks like some tests are failing CI but otherwise this looks like exactly what we need

I will be mostly away from my computer for the next couple of days. I'll be back on Monday and will be to to take a closer look, test it locally and get it merged.

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Thanks for adding test @Soptq.

It appears that the WETH withdrawal is being parsed by the Yearn parser, instead of the new one you created - we'll need to address that.

**Comment by Soptq** [comment]
> Thanks for adding test @Soptq.
> 
> It appears that the WETH withdrawal is being parsed by the Yearn parser, instead of the new one you created - we'll need to address that.

Fixed, it turns out to be a bug in the yearn parser, allowing the yearn parser to capture transactions even if they are not yearn-related.

File: node/packages/client/src/ethereum/parser/yearn.ts [hunk 1]
```diff
@@ -61,14 +61,19 @@ export class Parser implements SubParser {
...
 
-    if (txSigHash === yearnApprovalSigHash && decoded?.args._spender !== SHAPE_SHIFT_ROUTER_CONTRACT) return
-    if (txSigHash === shapeShiftDepositSigHash && receiveAddress !== SHAPE_SHIFT_ROUTER_CONTRACT) return
-    if (
-      txSigHash === yearnWithdrawSigHash &&
-      receiveAddress &&
-      !this.yearnTokenVaultAddresses?.includes(receiveAddress)
-    )
-      return
+    switch (txSigHash) {
```
Comment (0xApotheosis) [inline node/packages/client/src/ethereum/parser/yearn.ts:70]: Nice catch and refactor.

File: node/packages/client/src/ethereum/parser/yearn.ts [hunk 2]
```diff
@@ -61,14 +67,23 @@ export class Parser implements SubParser {
...
+      case yearnApprovalSigHash:
+        if (decoded?.args._spender !== SHAPE_SHIFT_ROUTER_CONTRACT) return
+        break
+      case shapeShiftDepositSigHash:
+        // here receiveAddress should be the address of the shapeShiftRouter contract
```
Comment (0xApotheosis) [inline node/packages/client/src/ethereum/parser/yearn.ts:75]: We can probably remove this comment, as its content is described by the code.

File: node/packages/client/src/ethereum/parser/yearn.ts [hunk 3]
```diff
@@ -61,14 +67,23 @@ export class Parser implements SubParser {
...
+        if (receiveAddress && receiveAddress !== SHAPE_SHIFT_ROUTER_CONTRACT) return
```
Comment (0xApotheosis) [inline node/packages/client/src/ethereum/parser/yearn.ts:76]:
```suggestion
        if (receiveAddress !== SHAPE_SHIFT_ROUTER_CONTRACT) return
```

**Review by 0xApotheosis — APPROVED** [review]
Some non-blocking comments, looking good!

**Comment by elmutt** [comment]
@Soptq @0xApotheosis 

I was just testing this PR locally and noticed some weth transactions unrelated to yearn are being processed by the yearn parser.

For example this TX:

https://etherscan.io/tx/0x29ed7ad51bbc568172d3d97bc79738a921e7dc669abbf15ab5198d97cc74614b

Gives:

```
{
  address: '0xE7F92E3d5FDe63C90A917e25854826873497ef3D',
  blockHash: '0x048e94784ee6376f08ce9e86e62ee304f3dfbb05b502f415ebc3ac090f939a1a',
  blockHeight: 14611926,
  blockTime: 1650322508,
  caip2: 'eip155:1',
  chain: 'ethereum',
  confirmations: 42907,
  fee: { caip19: 'eip155:1/slip44:60', value: '1087028000000000' },
  status: 'confirmed',
  tradeDetails: undefined,
  transfers: [
    {
      caip19: 'eip155:1/slip44:60',
      from: '0xE7F92E3d5FDe63C90A917e25854826873497ef3D',
      to: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2',
      type: 'send',
      value: '3264000000000000'
    }
  ],
  txid: '0x29ed7ad51bbc568172d3d97bc79738a921e7dc669abbf15ab5198d97cc74614b',
  data: { method: 'deposit', parser: 'yearn' }
}
```

Parser should be `weth` in this case because the tx has nothing to do with yearn and should be going through the weth parser.

**Review by elmutt — CHANGES_REQUESTED** [review]
See comment

**Comment by Soptq** [comment]
> https://etherscan.io/tx/0x29ed7ad51bbc568172d3d97bc79738a921e7dc669abbf15ab5198d97cc74614b

Hi @elmutt , I believe the transaction you tested is related to WETH. This transaction interacts with contract 0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2, which is the contract address of WETH. In other words, this transaction is indeed calling the `deposit` function of the WETH contract instead of yearn.

I think yearn probably proxy some WETH deposit/withdraw related operations to the WETH contract.

**Comment by Soptq** [comment]
Hi @elmutt , It seems I can't reproduce this problem. I added your provided transaction to the test:
- https://github.com/shapeshift/unchained/blob/87c1461465c45481224024b7680a14ec588b59b9/node/packages/client/src/ethereum/parser/__tests__/mockData/wethDeposit2.ts#L1-L35
-  https://github.com/shapeshift/unchained/blob/87c1461465c45481224024b7680a14ec588b59b9/node/packages/client/src/ethereum/parser/__tests__/ethereum.test.ts#L1318-L1372
and the test succeeded without problems as CI suggested.

@0xApotheosis Could you be able to reproduce this problem?

**Comment by elmutt** [comment]
@Soptq im looking into it a bit more now will let you know what I find

**Comment by elmutt** [comment]
> Hi @elmutt , It seems I can't reproduce this problem. I added your provided transaction to the test:
> 
> * https://github.com/shapeshift/unchained/blob/87c1461465c45481224024b7680a14ec588b59b9/node/packages/client/src/ethereum/parser/__tests__/mockData/wethDeposit2.ts#L1-L35
> * https://github.com/shapeshift/unchained/blob/87c1461465c45481224024b7680a14ec588b59b9/node/packages/client/src/ethereum/parser/__tests__/ethereum.test.ts#L1318-L1372
>   
>   and the test succeeded without problems as CI suggested.
> 
> @0xApotheosis Could you be able to reproduce this problem?

@Soptq I apologize, I was testing the wrong branch. On the correct branch the parser is correctly appearing as "weth"

I think I did find a small issue tho:

1) the "from" and "to" addresses on the weth transfers appear reversed. For example a weth "deposit" is creating new "weth" tokens for the user so the "to" address should be the users address and the "from" should be the contract address. (It looks reversed for withdraw as well, withdraw is a weth transfer from the user to the contract)

Otherwise this PR looks great and will get it merged & bounty paid after that last issue is fixed

**Comment by Soptq** [comment]
@elmutt Done! Thanks for your suggestions ;)

File: node/packages/client/src/ethereum/parser/weth.ts [hunk 4]
```diff
@@ -83,8 +83,8 @@ export class Parser implements SubParser {
           return [
             {
               type: TransferType.Receive,
```
Comment (elmutt) [inline node/packages/client/src/ethereum/parser/weth.ts:85]: Should be `TransferType.Send`

**Review by asamere — APPROVED** [review]

**Review by elmutt — APPROVED** [review]

**Comment by elmutt** [comment]
@Soptq Thanks for the the great work! Will get bounty paid out asap.

**Pull request merged**